providers/dns/designate/designate.go

```diff
@@ -0,0 +1,199 @@
+		PollingInterval:    env.GetOrDefaultSecond("DESIGNATE_POLLING_INTERVAL", 10*time.Second),
+		SequenceInterval:   env.GetOrDefaultSecond("DESIGNATE_SEQUENCE_INTERVAL", dns01.DefaultPropagationTimeout),
+	}
+}
+
+// DNSProvider describes a provider for Designate
+type DNSProvider struct {
+	config       *Config
+	client       *gophercloud.ServiceClient
+	dnsEntriesMu sync.Mutex
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for Designate.
+// Credentials must be passed in the environment variables:
+// OS_AUTH_URL, OS_USERNAME, OS_PASSWORD, OS_TENANT_NAME, OS_REGION_NAME.
+func NewDNSProvider() (*DNSProvider, error) {
+	_, err := env.Get("OS_AUTH_URL", "OS_USERNAME", "OS_PASSWORD", "OS_TENANT_NAME", "OS_REGION_NAME")
+	if err != nil {
+		return nil, errors.New("designate: missing Openstack credentials, please specify: OS_AUTH_URL, OS_USERNAME, OS_PASSWORD, OS_TENANT_NAME, OS_REGION_NAME")
+	}
+
+	opts, err := openstack.AuthOptionsFromEnv()
+	if err != nil {
+		return nil, fmt.Errorf("designate: %v", err)
+	}
+	config := NewDefaultConfig()
+	config.opts = opts
+
+	return NewDNSProviderConfig(config)
+}
+
+// NewDNSProviderConfig return a DNSProvider instance configured for Designate.
+func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
+	if config == nil {
+		return nil, errors.New("designate: the configuration of the DNS provider is nil")
+	}
+
+	provider, err := openstack.AuthenticatedClient(config.opts)
+	if err != nil {
+		return nil, fmt.Errorf("designate: failed to authenticate: %v", err)
+	}
+
+	dnsClient, err := openstack.NewDNSV2(provider, gophercloud.EndpointOpts{
+		Region: os.Getenv("OS_REGION_NAME"),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("designate: failed to get DNS provider: %v", err)
+	}
+
+	return &DNSProvider{client: dnsClient, config: config}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS propagation.
+// Adjusting here to cope with spikes in propagation times.
+func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return d.config.PropagationTimeout, d.config.PollingInterval
+}
+
+func getZoneID(client *gophercloud.ServiceClient, wanted string) (zoneID string, err error) {
+	allPages, err := zones.List(client, nil).AllPages()
+	if err != nil {
+		return
+	}
+	allZones, err := zones.ExtractZones(allPages)
+	if err != nil {
+		return
+	}
+
+	for _, zone := range allZones {
+		if zone.Name == wanted {
+			zoneID = zone.ID
+			return
+		}
+	}
+	return
+}
+
+func getRecordID(client *gophercloud.ServiceClient, zoneID string, wanted string) (recordID string, err error) {
+	allPages, err := recordsets.ListByZone(client, zoneID, nil).AllPages()
+	if err != nil {
+		return
+	}
+	allRecords, err := recordsets.ExtractRecordSets(allPages)
+	if err != nil {
+		return
+	}
+
+	for _, record := range allRecords {
+		if record.Name == wanted {
+			recordID = record.ID
+			return
+		}
+	}
+	return
```

Comment:
if the record ID is not found then an error must be returned.
Or you need to validate to recordID when you are using `getRecordID`